urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/Automaton.java

```diff
@@ -0,0 +1,1101 @@
+		}
+		return b.toString();
+	}
+
+	/**
+	 * Returns <a href="http://www.research.att.com/sw/tools/graphviz/" target="_top">Graphviz Dot</a> 
+	 * representation of this automaton.
+	 */
+	public String toDot() {
+		StringBuilder b = new StringBuilder("digraph Automaton {\n");
+		b.append("  rankdir = LR;\n");
+		Set<State> states = getStates();
+		setStateNumbers(states);
+		for (State s : states) {
+			b.append("  ").append(s.number);
+			if (s.accept)
+				b.append(" [shape=doublecircle,label=\"\"];\n");
+			else
+				b.append(" [shape=circle,label=\"\"];\n");
+			if (s == initial) {
+				b.append("  initial [shape=plaintext,label=\"\"];\n");
+				b.append("  initial -> ").append(s.number).append("\n");
+			}
+			for (Transition t : s.transitions) {
+				b.append("  ").append(s.number);
+				t.appendDot(b);
+			}
+		}
+		return b.append("}\n").toString();
+	}
+	
+	/**
+	 * Returns a clone of this automaton, expands if singleton.
+	 */
+	Automaton cloneExpanded() {
+		Automaton a = clone();
+		a.expandSingleton();
+		return a;
+	}
+
+	/**
+	 * Returns a clone of this automaton unless <code>allow_mutation</code> is set, expands if singleton.
+	 */
+	Automaton cloneExpandedIfRequired() {
+		if (allow_mutation) {
+			expandSingleton();
+			return this;
+		} else
+			return cloneExpanded();
+	}
+
+	/**
+	 * Returns a clone of this automaton.
+	 */
+	@Override
+	public Automaton clone() {
+		try {
+			Automaton a = (Automaton)super.clone();
+			if (!isSingleton()) {
+				HashMap<State, State> m = new HashMap<State, State>();
+				Set<State> states = getStates();
+				for (State s : states)
+					m.put(s, new State());
+				for (State s : states) {
+					State p = m.get(s);
+					p.accept = s.accept;
+					if (s == initial)
+						a.initial = p;
+					for (Transition t : s.transitions)
+						p.transitions.add(new Transition(t.min, t.max, m.get(t.to)));
+				}
+			}
+			return a;
+		} catch (CloneNotSupportedException e) {
+			throw new RuntimeException(e);
+		}
+	}
+	
+	/**
+	 * Returns a clone of this automaton, or this automaton itself if <code>allow_mutation</code> flag is set. 
+	 */
+	Automaton cloneIfRequired() {
+		if (allow_mutation)
+			return this;
+		else
+			return clone();
+	}
+	
+	/** 
+	 * Retrieves a serialized <code>Automaton</code> located by a URL.
+	 * @param url URL of serialized automaton
+	 * @exception IOException if input/output related exception occurs
+	 * @exception OptionalDataException if the data is not a serialized object
+	 * @exception InvalidClassException if the class serial number does not match
+	 * @exception ClassCastException if the data is not a serialized <code>Automaton</code>
+	 * @exception ClassNotFoundException if the class of the serialized object cannot be found
+	 */
+	public static Automaton load(URL url) throws IOException, OptionalDataException, ClassCastException, 
+	                                             ClassNotFoundException, InvalidClassException {
+		return load(url.openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when Automaton.load throws an exception. Ensure the object returned by *openStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).